### PR TITLE
Student: submit using submission link: give clearer indication of success #5979

### LINKED ISSUE #5979: Student: submit using submission link: give clearer indication of success
Some students say they are not sure if the submission went through. Perhaps we should use a modal dialog to inform the user of the submission status?

### PR DESCRIPTION
**Pull request by HirdayGupta**
Fixes #5979 

**Outline of solution**:
If either `studentFeedbackSubmissionEditPage` or the `instructorFeedbackSubmissionEditPage` load with a success status message, a modal is displayed in addition to the regular status message div to clearly indicate a successful submission. 

**Screenshot of modal:**
screen shot 2017-07-09 at 4 36 31 am

**Note:** I have noticed that the tick icon pushes the modal title to a new line in the modal header. This will be fixed in #7708.

**Question regarding approach (Senior dev input required):**
- As of now, this is only for unregistered users (who submit using submission link). Would it be necessary to add this in for registered students as well? (This will result in redirecting registered students to the feedback submission edit page as well after a successful submission instead of the current behavior of redirecting them to the student home page).

**Todo:**
- ~~Add test for visibility of modal.~~ Added this test in https://github.com/TEAMMATES/teammates/pull/7710/commits/08cc758242820c9f5daea0afa002bf28f8fb7429.

### CODE REVIEW

**Comment by damithc** [comment]
> As of now, this is only for unregistered users (who submit using submission link). Would it be necessary to add this in for registered students as well? (This will result in redirecting registered students to the feedback submission edit page as well after a successful submission instead of the current behavior of redirecting them to the student home page).

We can use the same approach for for both type of students.  There is a higher probability of registered students wanting to tweak the submission and resubmit rather than go back to home page.

**Comment by HirdayGupta** [comment]
@damithc changed the behavior to stay in edit page even after submitting the form for registered students also. Note that in the `createSpecificRedirectResult()` method responsible for deciding which page to redirect to, there was a provision to return to the edit page if there was an error. Since we are now returning to the same page whether or not there is an error, I have removed the if condition and modified the comment accordingly in d36a74a. Hoping this was the right call.

Ready for review. (Will update tests after review of feature).

**Comment by damithc** [comment]
> Hoping this was the right call.

Yes, that should be fine behavior-wise. I hope the model style is different when there is an error?

**Comment by HirdayGupta** [comment]
@damithc As far as I can see, no modal is displayed when there is an error. Just an `alert-danger` status message with the appropriate error message(s) (as set by `statusMessagesToUser`). The modal I have made in this PR will ONLY display if the responses had no errors and a success message was returned from the server. This is implemented by the following code:

```JavaScript
const SUCCESS_STATUS_MESSAGE = '.alert-success.statusMessage';
function hasSuccessMessage() {
    return $(SUCCESS_STATUS_MESSAGE).length;
}
```

The above code returns `1` only if there is a success message, and so the modal is only displayed if there is a success message.

**Comment by damithc** [comment]
> @damithc As far as I can see, no modal is displayed when there is an error
It's fine I guess.

**Comment by HirdayGupta** [comment]
@damithc I am now working on adding a test for modal and fixing the existing build. Let me know if any changes needed in code or current behaviour.

**Comment by damithc** [comment]
> @damithc I am now working on adding a test for modal and fixing the existing build.

Go ahead @HirdayGupta

**Comment by HirdayGupta** [comment]
**Summary of changes**
- Fixed build in commits d5e6658 and 1063a89 (In this, also added tests for modal visibility in `InstructorEditStudentFeedbackPageUiTest`)

- Added test for modal visiblity in `StudentFeedbackSubmitPageUiTest`

**Ready for review**

**Comment by HirdayGupta** [comment]
Dropping a note that the misaligned `success icon` problem will go away after #7774 is merged.

**Comment by HirdayGupta** [comment]
@whipermr5 Perhaps this can be changed to `s.ToReview`? #7774 Seems to be almost done and senoir devs can always wait before merging, but this PR has no direct relation to #7774.

File: src/main/java/teammates/ui/controller/StudentFeedbackSubmissionEditSaveAction.java [hunk 1]
```diff
@@ -75,20 +75,16 @@ protected RedirectResult createSpecificRedirectResult() {
...
         }
-        if (isError) {
-            // Return to student feedback submission edit page if there is an error and user is registered
-            RedirectResult result = createRedirectResult(Const.ActionURIs.STUDENT_FEEDBACK_SUBMISSION_EDIT_PAGE);
 
-            // Provide course id and session name for the redirected page
-            result.responseParams.put(Const.ParamsNames.COURSE_ID, student.course);
-            result.responseParams.put(Const.ParamsNames.FEEDBACK_SESSION_NAME,
-                                      getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME));
+        // Return to student home page if user is registered
```
Comment (leeyimin) [inline src/main/java/teammates/ui/controller/StudentFeedbackSubmissionEditSaveAction.java]: Update the comment.

File: src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java [hunk 2]
```diff
@@ -115,6 +117,7 @@ private void testDeleteResponse() {
                                           "student1InIESFPTCourse@gmail.tmt",
                                           "student1InIESFPTCourse@gmail.tmt");
         assertNull(fr);
+        submitPage.waitForAndDismissAlertModal();
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: Move this line to before line 110?

File: src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java [hunk 3]
```diff
@@ -75,6 +75,7 @@ private void testEditResponse() throws Exception {
                                           "student1InIESFPTCourse@gmail.tmt");
 
         assertEquals("<p>Good design</p>", fr.getResponseDetails().getAnswerString());
+        submitPage.waitForAndDismissAlertModal();
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: Move this line to before line 71?

**Comment by HirdayGupta** [comment]
@leeyimin All done. Ready for review! :)

File: src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java [hunk 4]
```diff
@@ -67,6 +67,7 @@ private void testEditResponse() throws Exception {
 
         submitPage.clickSubmitButton();
         submitPage.verifyStatus(Const.StatusMessages.FEEDBACK_RESPONSES_SAVED);
+        submitPage.waitForAndDismissAlertModal();
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: Could create a method in FeedbackSubmitPage to verify that the correct alert modal (message, style class) is created.

**Comment by HirdayGupta** [comment]
@leeyimin Added in the test for the correct modal. Ready for review! :)

**Comment by HirdayGupta** [comment]
Leaving a note that now that #7774 is merged, the modal is looking normal 🎉

Screenshot:

screen shot 2017-07-26 at 3 47 24 pm

**Comment by HirdayGupta** [comment]
@leeyimin My apologies, I didn't notice the failing test (Ran the wrong test locally before committing 😅) That being said, I am not entirely sure why the test is failing here. The exception is an `ElementNotVisibleException` but the element that it is trying to fill, the text box, is very much visible when I tried to debug using the `Thread.sleep()`. It seems strange that the same test method should work for the StudentFeedbackSubmitPageUiTest but not the InstructorEditStudentFeedbackPageUiTest.

**Comment by leeyimin** [comment]
@HirdayGupta add `waitForModalToDisappear` in your verify method.

**Comment by HirdayGupta** [comment]
Thanks @leeyimin! 
@thenaesh Awaiting review :)

**Comment by whipermr5** [comment]
@thenaesh Reminder

**Comment by HirdayGupta** [comment]
@thenaesh Reminder.

**Comment by HirdayGupta** [comment]
@thenaesh Please review

**Pull request merged**